src/Type/Php/CallUserFuncDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,33 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\Type;
+
+class CallUserFuncDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'call_user_func';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		$args = $functionCall->getArgs();
+		$defaultReturn = ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+
+		$callbackType = $scope->getType($args[0]->value);
+		if ($callbackType->isCallable()->yes()) {
+			return $scope->getType(new FuncCall($args[0]->value, array_slice($args, 1)));
```

Comment:
Does it matter if it's `FuncCall` when it's actually a static callback to a class method? Should we be returning `\PhpParser\Node\Expr\StaticCall` when appropriate?

Reply:
A unit test could tell. Will have a look

Reply:
This is fine. If you have `call_user_func(['Foo', 'bar'], 1, 2, 3)`, this will resolve to `(['Foo', 'bar'])(1, 2, 3)` which works.